Smoke test with mithril

ch1bo opened the pull request:
* Adds `--use-mithril` to `hydra-cluster` which makes it use `mithril-client` to download the cardano network state before starting `cardano-node`.

* Adds `mithril-client` to our development and CI shells.

* Adds integration tests to ensure that `mithril-client` is invoked correctly on all known networks.

A mithril-bootstrapped smoke test run on preview is here: https://github.com/input-output-hk/hydra/actions/runs/7540117340/job/20524247050

Not in scope / moved to red bin:

* Connection to cardano-node sometimes fails (too early?) https://github.com/input-output-hk/hydra/actions/runs/7541070906/job/20527001929

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Create downloadLatestSnapshotTo using mithril-client

Calls mithril-client to download and verify cardano-node db snapshots.
This is particularly useful for running scenarios on public (test)
networks.

ch1bo committed:
Add --use-mithril to hydra-cluster executable

This can be used to speed up smoke tests by downloading the latest
snapshot and have the cardano-node synchronize less.

ch1bo committed:
Update changelog

ch1bo committed:
Add use-mithril option to smoke test workflow

This will also delete the cardano-node db/ on the runner to download it
fresh from mithril.

ch1bo committed:
Add mithril-client to shells and test environments

v0d1ch commented on hydra-cluster/src/Hydra/Cluster/Mithril.hs:
```diff
@@ -0,0 +1,34 @@
+      , ["--download-dir", dir]
+      ]
+ where
+  genesisKeyURLForNetwork = \case
```
It would be nice if we can query these url's since I think I saw them change once.

v0d1ch commented on hydra-cluster/src/Hydra/Cluster/Options.hs:
```diff
@@ -4,26 +4,30 @@ import Data.ByteString.Char8 qualified as BSC
   deriving stock (Show)
 
--- TODO: Provide an option to use mithril aggregated snapshots to bootstrap the testnet
+data UseMithril = NotUseMithril | UseMithril
```
Maybe `NoMithril` or `DontUseMithril`?

ch1bo committed:
Add a MithrilLog tracer

This keeps the log output of tests clean and consistent with outputs of
cardano-node and hydra-node processes.

ch1bo committed:
Make downloadSnapshotTo tracing more robust

EOF errors are expected when reading all stdout (when the process exits)
and we can safely ignore them.

ch1bo committed:
Test only that mithril-client progresses to step 3

This makes the test quicker and asserts correct invocation.
